### ISSUE TITLE
'hdf5.h' file not found

### ISSUE THREAD

**Issue by mars0i** [vbrankov/hdf5-ocaml#9]
With OCaml 4.05.0+flambda, I get this error trying to install hdf5-ocaml with opam:
```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[ERROR] The compilation of hdf5 failed at "make".

#=== ERROR while compiling hdf5.0.1.4 =========================================#
# context      2.0.0~beta4 | darwin/x86_64 | ocaml-variants.4.05.0+flambda | https://opam.ocaml.org/2.0#2017-12-06 19:04
# path         ~/.opam/4.05.0+flambda/.opam-switch/build/hdf5.0.1.4
# command      make
# exit-code    2
# env-file     ~/.opam/log/hdf5-32892-05296d.env
# output-file  ~/.opam/log/hdf5-32892-05296d.out
### output ###
# [...]
# ocaml setup.ml -build
# ocamlfind ocamlopt -package unix -package ocamlbuild -linkpkg -package cppo_ocamlbuild myocamlbuild.ml /Users/marshall/.opam/4.05.0+flambda/lib/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
# /Users/marshall/.opam/4.05.0+flambda/bin/ocamlfind ocamlc -g -ccopt -I/usr/include/hdf5/serial -c lib/raw/h5a_stubs.c
# + /Users/marshall/.opam/4.05.0+flambda/bin/ocamlfind ocamlc -g -ccopt -I/usr/include/hdf5/serial -c lib/raw/h5a_stubs.c
# In file included from lib/raw/h5a_stubs.c:9:
# lib/raw/hdf5_caml.h:2:10: fatal error: 'hdf5.h' file not found
# #include "hdf5.h"
#          ^
# 1 error generated.
# Command exited with code 2.
# E: Failure("Command ''/Users/marshall/.opam/4.05.0+flambda/bin/ocamlbuild' lib/raw/libhdf5_raw_stubs.a lib/raw/dllhdf5_raw_stubs.so lib/raw/hdf5_raw.cma lib/raw/hdf5_raw.cmxa lib/raw/hdf5_raw.a lib/raw/hdf5_raw.cmxs lib/ppx/ppx_h5struct.native lib/caml/hdf5_caml.cma lib/caml/hdf5_caml.cmxa lib/caml/hdf5_caml.a lib/caml/hdf5_caml.cmxs -use-ocamlfind -plugin-tags 'package(cppo_ocamlbuild)' -tag[...]
# make: *** [build] Error 1
```

Maybe this is the same problem as #8?  It's not clear to me what it is I should install, in that case.  `libhdf5-dev` is an Ubuntu package, it seems.  I'm on MacOS.  I installed the only hdf5-related package (`hdf5`) that's available in homebrew.  I can install something else from source, but I'm not sure what's missing.

I'll try installing `hdf5-ocaml` from source later today.

**Comment by mars0i** [comment]
I should add that I had no problem on another MacOS machine.  Not sure yet what the difference is.  Theoretically, the machines configurations are the same, but they have different histories.

On the machine where I had the problem building with opam, I get the same error when I clone the repo and build from source.

**Comment by vbrankov** [comment]
The error almost certainly means that the development version of HDF5 library is not installed on the machine.  [depexts](https://github.com/vbrankov/hdf5-ocaml/blob/master/opam#L24) lists the packages that should be installed.  It shows that it needs `homebrew/science/hdf5` but I don't have a Mac to test it.  If you figure that out please let me know, I'll update `depexts`.

**Comment by mars0i** [comment]
I'm not sure what that path with `science` is (I'm not knowledgeable about homebrew), but I did have the hdf5 homebrew package installed, and deleted it and reinstalled.  I was able to get `hdf5-ocaml` built and installed by modifying `_oasis` in the cloned repo to look for the header and libs where I found them on my machine:
```
  CCOpt:    -I/usr/include/hdf5/serial -I/usr/local/include 
  CClib:    -L/usr/lib/x86_64-linux-gnu/hdf5/serial -L/usr/local/lib -lhdf5 -lhdf5_hl
``` 
This may be the wrong way to go about it, but it seemed to work, at least for `Hdf5_raw`.   I get an error for `Hdf5_caml`:
```
#require  "hdf5_caml";;
Error: The external function `caml_copy_string' is not available
```
Maybe that has to do with the `serial` lib?  Or a 4.06 vs 4.05 issue?  I haven't switched to 4.06 yet.

I have no idea (yet) why I didn't have any problem on my other Mac using opam.

I don't have the `serial` directory or library in `_oasis`.  Not sure what those are, but I only got warnings:
```
+ /Users/marshall/.opam/4.05.0+flambda/bin/ocamlfind ocamlmklib -o lib/raw/hdf5_raw_stubs -L/usr/lib/x86_64-linux-gnu/hdf5/serial -L/usr/local/lib -lhdf5 -lhdf5_hl lib/raw/h5a_stubs.o lib/raw/h5d_stubs.o lib/raw/h5f_stubs.o lib/raw/h5g_stubs.o lib/raw/h5i_stubs.o lib/raw/h5l_stubs.o lib/raw/h5lt_stubs.o lib/raw/h5o_stubs.o lib/raw/h5p_stubs.o lib/raw/h5r_stubs.o lib/raw/h5s_stubs.o lib/raw/h5_stubs.o lib/raw/h5tb_stubs.o lib/raw/h5t_stubs.o lib/raw/h5z_stubs.o lib/raw/hid_stubs.o
ld: warning: directory not found for option '-L/usr/lib/x86_64-linux-gnu/hdf5/serial'
+ /Users/marshall/.opam/4.05.0+flambda/bin/ocamlfind ocamlopt -shared -I lib/raw -I lib/raw lib/raw/hdf5_raw.cmxa lib/raw/hdf5_raw.cmx -o lib/raw/hdf5_raw.cmxs
ld: warning: directory not found for option '-L/usr/lib/x86_64-linux-gnu/hdf5/serial'
```
I'll keep playing with it.  I thought it might be worth submitting this interim report in case there is any obvious advice or diagnoses.   Thanks-

**Comment by vbrankov** [comment]
`caml_copy_string` is a part of the OCaml language library.  HDF5 should work on any OCaml starting with 4.02.  I really cannot tell why libraries cannot be automatically found on that machine, it doesn't seem to be either OCaml or HDF5 issue.

**Comment by mars0i** [comment]
Thanks.  That's weird about `caml_copy_string`, but good to have confirmation that that really should not be happening.  Maybe it's time for me to reinstall everything.  It might be a good time for me to switch to 4.06.

**Comment by mars0i** [comment]
btw I can't find a name with `copy_string` as part of it in the index of values for either OCaml 4.05 or 4.06:  http://caml.inria.fr/pub/docs/manual-ocaml/libref/index_values.html
I probably just don't know where to look.  I gather that this is not a regular OCaml function, but part of the interface to C.

**Comment by vbrankov** [comment]
It's here: https://caml.inria.fr/pub/docs/manual-ocaml-4.05/intfc.html

**Comment by mars0i** [comment]
Ah, thanks.

**Comment by mars0i** [comment]
I think that my little hack of the commandline flags in `_oasis` must have left something out.  I should probably start over regardless, since it works on the other Mac.

**Comment by mars0i** [comment]
The "`caml_copy_string' is not available" error occurs with hdf5 0.1.4 but not 0.1.3, fwiw.

**Comment by vbrankov** [comment]
0.1.3 has a bug with strings that `caml_copy_string` fixed in 0.1.4.

**Comment by mars0i** [comment]
I see.  I understand.  Thanks.  I'm working on what seems to be a glitch in my setup on the one machine.  Sometimes Apple makes things very difficult by trying to make them easy.  However, I don't think the default configuration in `_oasis` and the `opam` file are right for MacOS.  

opam file: `brew install homebrew/science/hdf5` reports that "homebrew/science/hdf5" is deprecated and that it's installing "hdf5".  I don't think this affects anything, however.

_oasis: I think that the include directories aren't right for MacOS, or at least for my version (10.11, El Capitan).  There's no `-I/usr/include/hdf5/serial` on either of my machines, even after I install `hdf5` with homebrew.  I've had success by adding "-I/usr/local/include" to the CCOpt line, and "-L/usr/local/lib" to the CClib line.

I'm moving issues about `Hdf5_caml` per se to a different issue.

**Comment by nilsbecker** [comment]
afaik everything homebrew installs ends up in `/usr/local/` so `/usr/include/` seems wrong to me as well.

**Comment by vbrankov** [comment]
mars0i what is the state of this?  Are there still open issues?

**Comment by vbrankov** [comment]
I changed the dependencies on MacOS so I'm closing as Travis compiles.

**Comment by mars0i** [comment]
@vbrankov, thanks again for the hdf5-ocaml project, and for coming back to this issue.  I haven't tried hdf5-ocaml since the period in which I submitted issues, and I'm not currently using OCaml.  (I'll come back to it, but it's not part of current projects.)  Whatever is in my past comments reflects my best understanding.  After that period, I stopped trying to use hdf5-ocaml, and worked up a small library based on `bin_prot`, which was good enough for my needs.  Because of my current projects, I can't do any experimentation with hdf5-ocaml right now.
